packages/labs/forms/src/form-associated.ts

```diff
@@ -0,0 +1,421 @@
+
+// #endregion
+
+// #region FormAssociated
+
+/**
+ * A mixin that makes a ReactiveElement into a form-associated custom element.
+ *
+ * - ARIA role: Sets this element's ElementInternals role to the value of the
+ *   static `role` property.
+ * - Form value: The value of the element is stored in the field decorated with
+ *   the `@formValue()` decorator. This value field can have any name, but the
+ *   type must be assignable to `string | File | FormData | null`.
+ * - Form state: The state of the element is stored in the field decorated with
+ *   the `@formState()` decorator. This state field can have any name, but the
+ *   type must be assignable to `string | File | FormData | null`. The state
+ *   field should be private, as it is only intended to be set by the
+ *   FormAssociated mixin. The state field should usually be a getter/setter
+ *   pair so that the form value can be derived from the state during form
+ *   state restoration.
+ * - Form reset: When the form is reset, the value field is set to its initial
+ *   value, and the state field is set to its initial state. This means that the
+ *   initial value and state will be stored for the lifetime of the element.
+ * - Form state restoration: When the form is restored, the mode can be either
+ *   'restore' or 'autocomplete'. In 'restore' mode, the state field is set to
+ *   the state setter, if present, is set to the state that was previously
+ *   passed to `internals.setFormValue()`. It's the state setter's
+ *   responsibility to update the value setter. If there is no state setter, the
+ *   value setter is updated directly. In 'autocomplete' mode, the value setter
+ *   is updated directly.
+ * - Form disabled: When the element is disabled, the `ariaDisabled` property of
+ *   the ElementInternals is set to "true". The element's disable state can be
+ *   checked with the `isDisabled()` function (an alias for
+ *   `element.matches(':disabled')`).
+ * - Validation: An element can implement the  `_getValidity()` method to
+ *   validate its state. This method is called automatically when the form value
+ *   changes, and can be called manually by the element.
+ */
+export const FormAssociated = <T extends Constructor<ReactiveElement>>(
+  base: T
+) => {
+  class C extends base implements FormAssociatedInterface {
+    static formAssociated = true;
+
+    // This must be a `super` call, not a `this` call!
+    #internals = super.attachInternals();
+    #attachInternalsCalled = false;
+
+    _getValidity?(): ValidityResult;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(...args: any[]) {
+      super(...args);
+      internalsCache.set(this, this.#internals);
+      this.#internals.role =
+        (this.constructor as FormAssociatedConstructor).role ?? null;
+    }
+
+    override attachInternals(): ElementInternals {
+      if (this.#attachInternalsCalled) {
+        throw new Error('attachInternals has already been called');
```

Comment:
Should this maybe just call `super.attachInternals()` to provoke the native error?